parallel-consumer-core/src/test/java/io/confluent/parallelconsumer/ParrallelEoSStreamProcessorPauseResumeTest.java

```diff
@@ -0,0 +1,246 @@
+package io.confluent.parallelconsumer;
+
+/*-
+ * Copyright (C) 2020-2022 Confluent, Inc.
+ */
+
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+import java.util.Optional;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.assertj.core.api.BDDAssertions;
+import org.awaitility.Awaitility;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Timeout;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.EnumSource;
+import org.slf4j.LoggerFactory;
+
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.Logger;
+import ch.qos.logback.classic.LoggerContext;
+import ch.qos.logback.classic.filter.LevelFilter;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.AppenderBase;
+import ch.qos.logback.core.spi.FilterReply;
+import io.confluent.parallelconsumer.ParallelConsumerOptions.CommitMode;
+import io.confluent.parallelconsumer.internal.AbstractParallelEoSStreamProcessor;
+import io.confluent.parallelconsumer.internal.BrokerPollSystem;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * Test for pause/resume feature of the parallel consumer (see {@code GH#193}).
+ *
+ * @author niels.oertel
+ */
+@Timeout(value = 10, unit = SECONDS)
+@Slf4j
+class ParrallelEoSStreamProcessorPauseResumeTest extends ParallelEoSStreamProcessorTestBase {
+
+    private static final AtomicLong MY_ID_GENERATOR = new AtomicLong();
+
+    private static final AtomicLong RECORD_BATCH_KEY_GENERATOR = new AtomicLong();
+
+    private ControlLoopTracker controlLoopTracker;
+
+    private static class TestUserFunction implements Consumer<ConsumerRecord<String, String>> {
+
+        AtomicLong numProcessedRecords = new AtomicLong();
+
+        @Override
+        public void accept(ConsumerRecord<String, String> t) {
+            numProcessedRecords.incrementAndGet();
+        }
+        
+        public void reset() {
+            numProcessedRecords.set(0L);
+        }
+    }
+
+    private ParallelConsumerOptions<String, String> getBaseOptions(final CommitMode commitMode) {
+        return ParallelConsumerOptions.<String, String>builder()
+                    .commitMode(commitMode)
+                    .consumer(consumerSpy)
+                    .maxConcurrency(3)
+                    .build();
+    }
+
+    private void addRecords(final int numRecords) {
+        long recordBatchKey = RECORD_BATCH_KEY_GENERATOR.incrementAndGet();
+        log.debug("Producing {} records with batch key {}.", numRecords, recordBatchKey);
+        for (int i = 0; i < numRecords; ++i) {
+            consumerSpy.addRecord(ktu.makeRecord("key-" + recordBatchKey + i, "v0-test-" + i));
+        }
+    }
+
+    /**
+     * Log message appender that monitors the entry messages of the two control
```

Comment:
use https://github.com/stefanbirkner/system-lambda#systemout-and-systemerr instead?

Reply:
I'm catching log events that are published on level TRACE which we don't want on the System.out.
Another option would be to add this hooks feature from `AbstractParallelEoSStreamProcessor` also to `BrokerPollSystem` but I don't really see any production use case for that and I'm not a big fan of adding such a test-only feature.

Reply:
As discussed - there is currently no other option to really count the loop cycles without either changing the log level to `TRACE` (would make test log output unreadable) or adding a hook to the `BrokerPollSystem` simularly to `AbstractParallelEoSStreamProcessor#controlLoopHooks` (feels rather wrong as it would be a test-only feature).